### PR TITLE
Fix: nondeterministic parameter validation (#2336)

### PR DESCRIPTION
**Pull request by amarjandu**
Author (before primary review)

- [x] PR title references issue
- [x] Title of main commit references issue
- [x] PR is linked to Zenhub issue
- [x] Added `DCP/1` label <sub>or this PR does not target an `hca/*` branch</sub>
- [x] Created issue to track porting these changes <sub>or this PR does not need porting</sub> 
- [x] Added `[r]` prefix at start of commit title <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label <sub>or this PR does not require reindexing</sub>
- [x] Freebies are blocked on this PR <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in chain <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR <sub>or this PR is not chained to another PR</sub>
- [x] Ran `make requirements_update` <sub>or this PR does not change requirements*.txt</sub>
- [x] Mentioned `make requirements` in UPGRADING.rst <sub>or this PR does not change requirements*.txt</sub>
- [x] Documented upgrade of personal deployments in UPGRADING.rst <sub>or this PR does not require upgrading</sub>
- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>

Primary reviewer (before pushing merge commit)

- [x] Checked `reindex` label and `[r]` prefix of commit title
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Build passes in `sandbox` <sub>or added `no sandbox` label</sub>
- [x] Reindexed `sandbox` <sub>or this PR does not require reindexing</sub>
- [x] Added PR reference to merge commit
- [x] Moved linked issue to Merged
- [x] Pushed merge commit to Github

Primary reviewer (after pushing merge commit)

- [ ] Moved freebies to Merged column <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened chain <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [ ] Commented on demo expectations <sub>or labeled as `no demo`</sub>
- [ ] Pushed merge commit to Gitlab <sub>or this changes can be pushed later, together with another PR</sub>
- [ ] Reindexed `dev` <sub>or this PR does not require reindexing</sub>
- [ ] Deleted PR branch from Github and Gitlab
- [ ] Unassign reviewer from PR

### CODE REVIEW

**Comment by amarjandu** [comment]
Created this PR to see if the issue can be reproduced on develop within Travis.

**Comment by amarjandu** [comment]
Was not able to replicate the test failures in the same way that #1465 was failing, but the change in the ordering for the query parameters can be observed when debugging tests.

**Review by achave11 — APPROVED** [review]
Looks good to me!

File: test/service/test_request_validation.py [hunk 1]
```diff
@@ -106,7 +106,7 @@ def test_bad_sort_facet_and_filter_facet_of_sample(self):
         }
         response = requests.get(url, params=params)
         self.assertEqual(400, response.status_code, response.json())
-        self.assertTrue(response.json() in [self.facet_message, self.facet_message])
```
Comment (hannes-ucsc) [inline test/service/test_request_validation.py:109]: @achave11 FYI

**Review by hannes-ucsc — COMMENTED** [review]
There was still some non-determinism left. Please review.

**Comment by hannes-ucsc** [comment]
Review my changes, that is.

**Comment by amarjandu** [comment]
@hannes-ucsc there are many short-circuits to a BadRequestError that escapes the collection of invalid_params. 

For some query parameters:

```python
        params = {
            'order': 'up',
            'size': '23',
            'sort': 'bad-facet',
            'filters': json.dumps({'organ': {'is': ['fake-val2']}}),
        }
```

In the suggested changes, the invalid `order` parameter would be collected however it will not be returned to the caller because of the validate_size BadRequestError that will occur.

Now if we start catching all the short circuits we could return an error message as seen in [RFC7807](https://tools.ietf.org/html/rfc7807#section-3)

This would allow us to consistently format the output that is sent back to the users, ensure that all the parameters are inspected, and provide insightful responses. 

The response could look something along the lines of: 

```js
{
    'invalid_parameters': [
        {
            'name': 'order',
            'value': 'up',
            'error': 'invalid'
        },
        {
            'name': 'size',
            'value': '23',
            'error': 'type'
        },
        {
            'name': 'sort',
            'value': 'bad-facet',
            'error': 'invalid'
        },
        { 
            'name': 'filter',
            'value': {'organ': {'is': ['fake-val2']}},
            'error': 'unknown relation value'
        },
    ]
}
```

File: lambdas/service/app.py [hunk 2]
```diff
@@ -827,33 +827,34 @@ def validate_params(query_params: Mapping[str, str],
...
-        except (TypeError, ValueError, RequirementError):
-            raise BadRequestError(msg=f'Invalid input value for `{param_name}`')
+            validator = validators[param_name]
+        except KeyError:
+            pass
+        else:
+            try:
+                validator(param_value)
+            except (TypeError, ValueError, RequirementError):
+                raise BadRequestError(msg=f'Invalid input value for `{param_name}`')
```
Comment (hannes-ucsc) [inline lambdas/service/app.py:857]:
```suggestion
                raise BadRequestError(msg=f'Invalid value for `{param_name}`')
```

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Comment by hannes-ucsc** [comment]
No bump.

**Review by hannes-ucsc — APPROVED** [review]

**Comment by hannes-ucsc** [comment]
For demo, show that validation still works by making requests with a missing, an extra and an invalid parameter, each. Show that multiples of those work, too.

**Pull request merged**